Fix findAndCountAll issues

sushantdhiman opened the pull request:
Fixed the `subQuery` generation which cause nested `includes` to be applied on main query. It also add other tests and fixes achieved due to this fix.

Fixes #4191, Fixes #3009, Fixes #5445
Closes #5530, Closes #5446, Closes #4378

May be it will close more issues, I will check and include tests in further PRs

sushantdhiman commented:
@mickhansen @janmeier  Please have a look on this one when you get chance

mickhansen commented on changelog.md:
```diff
@@ -1,4 +1,7 @@
 # Future
+- [Fixed] `findAndCountAll` generates wrong SQL query [#5445](https://github.com/sequelize/sequelize/issues/5445)
+- [Fixed] `count` methods pollute the options.includes [#4191](https://github.com/sequelize/sequelize/issues/4191)
```
I don't see a change for this, have you just verified that it does indeed work?

sushantdhiman replied:
#4191 , I have added a test for that https://github.com/sushantdhiman/sequelize/commit/3f15e064f274c2219c982f4b7fffe757956f6ac0

mickhansen replied:
Right, more tests are awesome! Just wanted to check if it was already broken, and doesn't seem like it was

sushantdhiman replied:
Yes its want broken because someone fixed .findAll method, But still there was no test. So its a verification only, it was already fixed

mickhansen commented on lib/model.js:
```diff
@@ -484,6 +484,7 @@ var validateIncludedElements = function(options, tableNames) {
         include.subQuery = false;
       } else {
         include.subQueryFilter = false;
+        include.subQuery = include.hasParentRequired && (include.hasDuplicating || include.hasRequired);
```
This would overwrite the line above that supports an already set `include.subQuery`.
Simply prefix the hasParentRequired check with `include.subQuery || ` and it should be ok.

sushantdhiman commented:
@mickhansen I have prefixed the condition, Can you somehow start the CI build, My local test run was ok.

mickhansen commented:
No CI active for this PR, very odd.

sushantdhiman commented:
Yeah Travis didn't started any CI because last commit had `[ci skip]` flag, I have renamed the commit. CI will start now.

sushantdhiman commented:
@mickhansen current CI has stuck, Can you please restart

mickhansen commented:
Please re-open and i'll make sure it gets a thorough review.

mickhansen commented:
Still no CI, very odd, no idea how to fix.

mickhansen commented on test/integration/model/findAll/group.test.js:
```diff
@@ -0,0 +1,57 @@
+      });
+    });
+  });
+});
```
What are you verifying in this test?

sushantdhiman replied:
Its verification for https://github.com/sequelize/sequelize/issues/3009 , standalone commit https://github.com/sequelize/sequelize/pull/5534/commits/5d0f28a38cde3ecb434963264aa05858d89509cc which add this test. While browsing through count and find all issues I came across this. Issue was already fixed so I included it in this PR

mickhansen commented on test/integration/include/findAndCountAll.test.js:
```diff
@@ -207,5 +207,78 @@ describe(Support.getTestDialectTeaser('Include'), function() {
+        });
+      }).then(function(result) {
+          expect(result.count).to.equal(2);
+          expect(result.rows.length).to.equal(2);
```
Shouldn't this equal 1? There's only one task that matches the combination of A and user-name-2.
Ideally when testing findAndCountAll it would be great if the LIMIT was lower than rows so we can verify that both parts work correctly.

sushantdhiman replied:
The result will match to 2, because for `user-name-2` there is one project `A` with 2 tasks `testing` and `purpose`.

Yes `findAndCountAll` test will be much better with LIMIT set to `1` so we can match correct count and records. I will modify it

mickhansen replied:
Ah right you're counting tasks, my bad.

mickhansen commented on lib/model.js:
```diff
@@ -484,6 +484,7 @@ var validateIncludedElements = function(options, tableNames) {
         include.subQuery = false;
       } else {
         include.subQueryFilter = false;
+        include.subQuery = include.subQuery || include.hasParentRequired && (include.hasDuplicating || include.hasRequired);
```
We need to add the include to the subquery if it's parent is in the subquery and itself or it's children is required.
Do we care about `hasDuplicated` here?

sushantdhiman replied:
I just need the `true` value here because its causing the nested queries to go with LEFT OUTER JOIN even if they should be part of INNER JOIN.

If we consider we just need to respect `hasParentRequired` and `hasRequired` flags only, so there is no particular role for `hasDuplicating` here. I think I saw some test failing because of that. But that time code didn't included the `hasParentRequired` which I added later on after your suggestion. I will repeat the tests to check if we really need the `hasDuplicating` now.

mickhansen replied:
I suggest `hasParentRequired`?
I can't really see why it's needed, whether or not an include parent is in the subquery doesn't matter unless the specific include needs to be in the subquery (at which point the parent should be forced, but that happens at the parent level).

sushantdhiman replied:
Ah you suggested to add `include.subQuery`, before adding that when I was looking for cause of the subQuery inclusion I found that `hasDuplicating` was also needed as deciding a factor. 

Now since logic has been changed, I'll need to check if we still need the `hasDuplicating` or any factor other than `include.subQuery`

sushantdhiman commented:
Hi @mickhansen , changed the condition and removed `hasDuplicating` as its not required now, rebased code, [CI link here](https://travis-ci.org/sequelize/sequelize/builds/116824746)

mickhansen commented:
Looks great @sushantdhiman, could you change the test to use a lower limit than the count?

sushantdhiman commented:
Let me know if this is correct behavior @mickhansen 

SELECT  `Task`.*, 

`Project.User`.`id` AS `Project.User.id`, `Project.User`.`name` AS `Project.User.name`, `Project.User`.`createdAt` AS `Project.User.createdAt`, `Project.User`.`updatedAt` AS `Project.User.updatedAt`, 

`Tags`.`id` AS `Tags.id`, `Tags`.`name` AS `Tags.name`, `Tags`.`createdAt` AS `Tags.createdAt`, `Tags`.`updatedAt` AS `Tags.updatedAt`, 

`Tags.TaskTag`.`id` AS `Tags.TaskTag.id`, `Tags.TaskTag`.`name` AS `Tags.TaskTag.name`, `Tags.TaskTag`.`createdAt` AS `Tags.TaskTag.createdAt`, `Tags.TaskTag`.`updatedAt` AS `Tags.TaskTag.updatedAt`, `Tags.TaskTag`.`TaskId` AS `Tags.TaskTag.TaskId`,

 `Tags.TaskTag`.`TagId` AS `Tags.TaskTag.TagId` FROM (
SELECT `Task`.`id`, `Task`.`name`, `Task`.`createdAt`, `Task`.`updatedAt`, `Task`.`ProjectId`, `Project`.`id` AS `Project.id`, `Project`.`m` AS `Project.m`, `Project`.`createdAt` AS `Project.createdAt`, `Project`.`updatedAt` AS `Project.updatedAt`, `Project`.`UserId` AS `Project.UserId` FROM `Tasks` AS `Task` INNER JOIN `Projects` AS `Project` ON `Task`.`ProjectId` = `Project`.`id` AND (`Project`.`m` = 'A') ORDER BY `Task`.`id` ASC LIMIT 1) 
**`--- LIMIT APPLIED IN SUB QUERY`**

AS `Task`  

INNER JOIN `Users` AS `Project.User` ON `Project.UserId` = `Project.User`.`id` AND (`Project.User`.`name` = 'user-name-2') 

LEFT OUTER JOIN (`TaskTags` AS `Tags.TaskTag` INNER JOIN `Tags` AS `Tags` ON `Tags`.`id` = `Tags.TaskTag`.`TagId`) ON `Task`.`id` = `Tags.TaskTag`.`TaskId` 

ORDER BY `Task`.`id` ASC;  **`--- LIMIT SHOULD BE HERE ?`**

This means first we filter by LIMIT then apply JOINS on filtered IDs, Is this correct?

mickhansen commented:
The limit has to be on the inner query so it limits the rows we want to limit and not the cartesian product.
If a include has a where/required and it's not duplicating it can be moved to the subquery, if it's duplicating it needs to be on the outer query and add an additional subquery as filter to the subquery.

sushantdhiman commented:
@mickhansen If thats example is correct way then great, I have changed the `limit` to `1`.

I would like to mention that when we change order to `ASC` then test will fail, because we do filtering in sub-query first, this will get first i.e. `task 1` (because limit is `1`), then we check for `project = "A" and user = "user-name-2" and task IN (1)`.

Since we have limited task id to `1`, the main result is empty because required task ids `3, 4` (for project `A` and user `user-name-2`) were left out during sub query filter.

mickhansen commented:
@sushantdhiman Ordering shouldn't matter, a bug if it does.

sushantdhiman commented:
Thanks
